x/datapool/keeper/pool_distribute.go

```diff
@@ -0,0 +1,226 @@
+	key := types.GetKeyPrefixSalesHistory(
+		salesHistory.PoolId,
+		salesHistory.Round,
+		salesHistory.SellerAddress,
+	)
+	store := ctx.KVStore(k.storeKey)
+	store.Set(key, k.cdc.MustMarshalBinaryLengthPrefixed(salesHistory))
+}
+
+// GetSalesHistory returns the sales history. If there is no value, it responds nil.
+func (k Keeper) GetSalesHistory(ctx sdk.Context, poolID, round uint64, sellerAddress string) *types.SalesHistory {
+	key := types.GetKeyPrefixSalesHistory(poolID, round, sellerAddress)
+	store := ctx.KVStore(k.storeKey)
+	if !store.Has(key) {
+		return nil
+	}
+	bz := store.Get(key)
+	var salesHistory types.SalesHistory
+	k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &salesHistory)
+	return &salesHistory
+}
+
+func (k Keeper) ListSalesHistory(ctx sdk.Context, poolID, round uint64) []*types.SalesHistory {
+	key := types.GetKeyPrefixSalesHistories(poolID, round)
+	store := ctx.KVStore(k.storeKey)
+	iter := sdk.KVStorePrefixIterator(store, key)
+	var histories []*types.SalesHistory
+	for ; iter.Valid(); iter.Next() {
+		history := &types.SalesHistory{}
+		bz := iter.Value()
+		k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, history)
+		histories = append(histories, history)
+	}
+	return histories
+}
+
+// SetInstantRevenueDistribute stores the poolID to which the revenue should be distributed immediately.
+func (k Keeper) SetInstantRevenueDistribute(ctx sdk.Context, instantRevenueDistribute *types.InstantRevenueDistribute) {
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(instantRevenueDistribute)
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.KeyPrefixInstantRevenueDistribute, bz)
+}
+
+func (k Keeper) GetInstantRevenueDistribute(ctx sdk.Context) *types.InstantRevenueDistribute {
+	store := ctx.KVStore(k.storeKey)
+	if !store.Has(types.KeyPrefixInstantRevenueDistribute) {
+		return &types.InstantRevenueDistribute{}
+	}
+	bz := store.Get(types.KeyPrefixInstantRevenueDistribute)
+	var delayedRevenueDistribute types.InstantRevenueDistribute
+	k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &delayedRevenueDistribute)
+	return &delayedRevenueDistribute
+}
+
+// addInstantRevenueDistribute adds the poolID to distribute. If there are duplicate poolIDs, they are not added.
+func (k Keeper) addInstantRevenueDistribute(ctx sdk.Context, poolID uint64) {
+	instantRevenueDistribute := k.GetInstantRevenueDistribute(ctx)
+	instantRevenueDistribute.AppendPoolID(poolID)
+	k.SetInstantRevenueDistribute(ctx, instantRevenueDistribute)
+}
+
+func (k Keeper) DistributeRevenuePools(ctx sdk.Context) error {
+	instantRevenueDistribute := k.GetInstantRevenueDistribute(ctx)
+
+	if instantRevenueDistribute.IsEmpty() {
+		return nil
+	}
+
+	// This lastIndex is to execute the following iteration in the next block if it does not end in one block.
+	// But We're currently processing it all in one block. :)
+	lastIndex := 0
+	// TODO We need to think about how to deal with the pool where distribution has failed.
```

Comment:
From my understanding, if the BeginBlocker panics, the state machine is halted. So, we don't need to worry about how to rollback operations that succeeded partially.
But, we need to test it.

Reply:
Yes, we will proceed as you understands.
As you said, we need to test this.